lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1172 @@
+	}
+	
+	// Useful method for checking which branches(T/F) are saved in the current lattice (rlc.constant or even a value)
+	const std::pair<bool,bool> ConstraintsLattice::branchesContainedInLattice() const{
+		bool can_true=false;
+		bool can_false=false;
+
+		for(auto pair : this->ranges){
+			if(pair.first.second)
+				can_true=true;
+			else can_false=true;
+		}
+
+		return std::make_pair(can_true,can_false);
+	}
+
+	// Method for checking if the value in input is contained in both branches in the lattice
+	const std::pair<bool,bool> ConstraintsLattice::isValuePresentInLattice(const mlir::Value& value) const{
+		bool can_true=false;
+		bool can_false=false;
+
+		for(auto pair : this->ranges){
+			if(pair.first.first==value){
+				if(pair.first.second)
+					can_true=true;
+				else can_false=true;
+			}
+		}
+
+		return std::make_pair(can_true,can_false);
+	}
+
+	// Method for checking if there is a constant in the lattice
+	const std::pair<bool,bool> ConstraintsLattice::isConstPresentInLattice() const{
+		bool can_true=false;
+		bool can_false=false;
+
+		for(auto pair : this->ranges){
+			// Sanity check as the DefiningOp of function arguments is nullptr
+			if(not (pair.first.first.getDefiningOp()==nullptr) and
+				mlir::isa<mlir::rlc::Constant>(pair.first.first.getDefiningOp())){
+				if(pair.first.second)
+					can_true=true;
+				else can_false=true;
+			}
+		}
+
+		return std::make_pair(can_true,can_false);
+	}
+
+	// Useful case for branch checking
+	bool ConstraintsLattice::isOnlyElementInToCheckLattice(const mlir::Value& value) const{
+		return to_check.contains(value) and this->ranges.empty();
+	}
+
+	// Useful case for branch checking
+	bool ConstraintsLattice::isOnlyToCheckPresentInLattice(const mlir::Value& value) const{
+		
+		for(auto pair : this->ranges){
+			if(pair.first.first==value)
+				return false;
+		}
+		return not this->to_check.empty();
+	}
+
+	// Useful case for branch checking
+	bool ConstraintsLattice::isValueOfTypePresentInLattice(bool type) const{
+
+		for(auto pair : this->ranges){
+			if(pair.first.second==type)
+				return true;
+		}
+		return false;
+	}
+
+	// Useful case for branch checking
+	bool ConstraintsLattice::rangesEmptyAndToCheckOneElement() const{
+		return this->ranges.empty() and this->to_check.size()==1;
+	}
+
+	// Useful method
+	bool ConstraintsLattice::rangesEmpty() const{
+		return this->ranges.empty();
+	}
+
+	// Useful method
+	bool ConstraintsLattice::toCheckEmpty() const{
+		return this->to_check.empty();
+	}
+
+	// Useful method
+	bool ConstraintsLattice::isInToCheck(const mlir::Value& value) const{
+		auto found=this->to_check.find(value);
+		return found!=this->to_check.end();
+	}
+
+	// Method for returning the branches values for the rlc.builtin_assign operation
+	const std::tuple<bool,bool,bool,bool> ConstraintsLattice::GetBranchesBuiltinAssign(mlir::Operation* statement, mlir::rlc::ConstraintsAnalysis* analysis, const mlir::Value& value){
```

Comment:
i understand that this method is important, but this is fairly magical, it is handling builtin assign ad hoc to work around the issues it generates. if possible i would want to lose special status and express this trough the interface